From 9443ef482c9d07600b3a60a1c2040463b38e6147 Mon Sep 17 00:00:00 2001
From: Samin Guo <samin.guo@starfivetech.com>
Date: Fri, 8 Jan 2021 03:11:04 +0800
Subject: [PATCH 02/29] drivers/tty/serial/8250: update driver for VIC7100

---
 drivers/tty/serial/8250/8250_port.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/tty/serial/8250/8250_port.c b/drivers/tty/serial/8250/8250_port.c
index d45dab1ab316..d05d3439789e 100644
--- a/drivers/tty/serial/8250/8250_port.c
+++ b/drivers/tty/serial/8250/8250_port.c
@@ -73,8 +73,16 @@ static const struct serial8250_config uart_config[] = {
 	},
 	[PORT_16550] = {
 		.name		= "16550",
+#ifdef CONFIG_SOC_STARFIVE_VIC7100
+		.fifo_size	= 16,
+		.tx_loadsz	= 16,
+		.fcr		= UART_FCR_ENABLE_FIFO | UART_FCR_R_TRIG_00,
+		.rxtrig_bytes	= {1, 4, 8, 14},
+		.flags		= UART_CAP_FIFO,
+#else
 		.fifo_size	= 1,
 		.tx_loadsz	= 1,
+#endif
 	},
 	[PORT_16550A] = {
 		.name		= "16550A",
-- 
2.30.0

