From 8778e98f2cde3de0155aea8fb5ccc24332d76426 Mon Sep 17 00:00:00 2001
From: Tom <support@vamrs.com>
Date: Fri, 8 Jan 2021 02:59:40 +0800
Subject: [PATCH 30/56] riscv: Add StarFive JH7100 support

---
 arch/riscv/Kconfig.socs | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/arch/riscv/Kconfig.socs b/arch/riscv/Kconfig.socs
index 30676ebb16eb..35b8ccf4e84b 100644
--- a/arch/riscv/Kconfig.socs
+++ b/arch/riscv/Kconfig.socs
@@ -19,6 +19,18 @@ config SOC_SIFIVE
 	help
 	  This enables support for SiFive SoC platform hardware.
 
+config SOC_STARFIVE_VIC7100
+	bool "StarFive VIC7100 SoC"
+	select SOC_SIFIVE
+	select OF_RESERVED_MEM
+	select SIFIVE_L2
+	select SIFIVE_L2_FLUSH
+	select DW_AXI_DMAC_STARFIVE
+	select PINCTRL
+	select PINCTRL_STARFIVE
+	help
+	  This enables support for StarFive VIC7100 SoC Platform Hardware.
+
 config SOC_VIRT
 	bool "QEMU Virt Machine"
 	select CLINT_TIMER if RISCV_M_MODE
-- 
2.30.2

