From 2c05d96f361f0077337b2ce21aa0d5ac459054a6 Mon Sep 17 00:00:00 2001
From: Stephen L Arnold <nerdboy@gentoo.org>
Date: Mon, 17 May 2021 14:44:05 -0700
Subject: [PATCH] drivers: hwmon: sfctmp: remove scaling, return value in
 millideg C

Signed-off-by: Stephen L Arnold <nerdboy@gentoo.org>
---
 drivers/hwmon/sfctmp.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/hwmon/sfctmp.c b/drivers/hwmon/sfctmp.c
index 901f92f8dd2d..b253c2b375c6 100644
--- a/drivers/hwmon/sfctmp.c
+++ b/drivers/hwmon/sfctmp.c
@@ -61,13 +61,13 @@ struct sfc_temp{
 
 static ssize_t sfctmp_get_temp(struct device *dev, struct device_attribute *devattr,char *buf)
 {
-	long temp,temp_z,temp_x;
-	const long Y100 = 23750, K100 = 8110,Z100 = 409400;
+	long temp;
+	const long Y100 = 23750, K100 = 8110, Z100 = 409400;
 
 	temp  = ((long)s_temp_sensor_dout*100)*Y100/Z100-K100;
-	temp_z = temp/100;
-	temp_x = temp%100;
-	return	sprintf(buf, "%ld.%ld\n", temp_z,temp_x);
+
+	// multiply by 10 so we can return a value in millideg C
+	return	sprintf(buf, "%ld\n", temp*10);
 }
 
 static SENSOR_DEVICE_ATTR(temp1_input, S_IRUGO, sfctmp_get_temp, NULL, 0);
@@ -76,7 +76,7 @@ static ssize_t sfctmp_temp_show(void)
 {
 	char buf[10];
 	sfctmp_get_temp(NULL, NULL, buf);
-	printk("temp(c): %s",buf);
+	printk("temp(mdeg C): %s",buf);
 	return 0;
 }
 
@@ -208,4 +208,4 @@ module_platform_driver(sfc_temp_sensor_driver);
 
 MODULE_AUTHOR("samin.guo");
 MODULE_DESCRIPTION("SFC temperature sensor driver");
-MODULE_LICENSE("GPL");
\ No newline at end of file
+MODULE_LICENSE("GPL");
-- 
2.30.0

