From 200db02037031341e1df85c68eb984c2f1a5f214 Mon Sep 17 00:00:00 2001
From: Atish Patra <atish.patra@wdc.com>
Date: Sat, 12 Jun 2021 11:02:41 -0700
Subject: [PATCH 40/42] Remove cache flush code from fbcon driver

Signed-off-by: Atish Patra <atish.patra@wdc.com>
---
 drivers/video/fbdev/core/fbcon.c | 24 ------------------------
 1 file changed, 24 deletions(-)

diff --git a/drivers/video/fbdev/core/fbcon.c b/drivers/video/fbdev/core/fbcon.c
index e3138bfe279d..1c40d64682ef 100644
--- a/drivers/video/fbdev/core/fbcon.c
+++ b/drivers/video/fbdev/core/fbcon.c
@@ -82,11 +82,6 @@
 
 #ifdef CONFIG_SOC_STARFIVE_VIC7100
 #include <video/starfive_fb.h>
-#include <soc/starfive/vic7100.h>
-static inline void fbcon_flush_dcache(unsigned long start, unsigned long len)
-{
-	starfive_flush_dcache(_ALIGN_DOWN(start, 64), len + start % 64);
-}
 #endif
 
 /*
@@ -1697,9 +1692,6 @@ static bool fbcon_scroll(struct vc_data *vc, unsigned int t, unsigned int b,
 	struct fb_info *info = registered_fb[con2fb_map[vc->vc_num]];
 	struct fbcon_display *p = &fb_display[vc->vc_num];
 	int scroll_partial = info->flags & FBINFO_PARTIAL_PAN_OK;
-#ifdef CONFIG_SOC_STARFIVE_VIC7100
-	struct sf_fb_data *sf_dev = container_of(info, struct sf_fb_data, fb);
-#endif
 	if (fbcon_is_inactive(vc, info))
 		return true;
 
@@ -1727,10 +1719,6 @@ static bool fbcon_scroll(struct vc_data *vc, unsigned int t, unsigned int b,
 							(b - count)),
 				    vc->vc_video_erase_char,
 				    vc->vc_size_row * count);
-#ifdef CONFIG_SOC_STARFIVE_VIC7100
-			fbcon_flush_dcache(sf_dev->fb.fix.smem_start,
-					   sf_dev->fb.screen_size);
-#endif
 			return true;
 
 		case SCROLL_WRAP_MOVE:
@@ -1802,10 +1790,6 @@ static bool fbcon_scroll(struct vc_data *vc, unsigned int t, unsigned int b,
 							(b - count)),
 				    vc->vc_video_erase_char,
 				    vc->vc_size_row * count);
-#ifdef CONFIG_SOC_STARFIVE_VIC7100
-			fbcon_flush_dcache(sf_dev->fb.fix.smem_start,
-					   sf_dev->fb.screen_size);
-#endif
 			return true;
 		}
 		break;
@@ -1825,10 +1809,6 @@ static bool fbcon_scroll(struct vc_data *vc, unsigned int t, unsigned int b,
 							t),
 				    vc->vc_video_erase_char,
 				    vc->vc_size_row * count);
-#ifdef CONFIG_SOC_STARFIVE_VIC7100
-			fbcon_flush_dcache(sf_dev->fb.fix.smem_start,
-					   sf_dev->fb.screen_size);
-#endif
 			return true;
 
 		case SCROLL_WRAP_MOVE:
@@ -1898,10 +1878,6 @@ static bool fbcon_scroll(struct vc_data *vc, unsigned int t, unsigned int b,
 							t),
 				    vc->vc_video_erase_char,
 				    vc->vc_size_row * count);
-#ifdef CONFIG_SOC_STARFIVE_VIC7100
-			fbcon_flush_dcache(sf_dev->fb.fix.smem_start,
-					   sf_dev->fb.screen_size);
-#endif
 			return true;
 		}
 	}
-- 
2.31.1

