From a13941ca02dbc4dd97366ebccc426d4fef5a516a Mon Sep 17 00:00:00 2001
From: Atish Patra <atish.patra@wdc.com>
Date: Sat, 12 Jun 2021 16:52:26 -0700
Subject: [PATCH 42/42] sifive/sifive_l2_cache: Align the address to cache line

Signed-off-by: Atish Patra <atish.patra@wdc.com>
---
 drivers/soc/sifive/sifive_l2_cache.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/soc/sifive/sifive_l2_cache.c b/drivers/soc/sifive/sifive_l2_cache.c
index 626b664547e5..58983da9ac97 100644
--- a/drivers/soc/sifive/sifive_l2_cache.c
+++ b/drivers/soc/sifive/sifive_l2_cache.c
@@ -139,6 +139,9 @@ void sifive_l2_flush64_range(unsigned long start, unsigned long len)
 		return;
 	}
 
+	start = ALIGN_DOWN(start, 64);
+	len = len + (start % 64);
+
 	/* make sure the address is in the range */
 	if(start < CONFIG_SIFIVE_L2_FLUSH_START ||
 	   (start + len) > (CONFIG_SIFIVE_L2_FLUSH_START +
-- 
2.31.1

