From a65dd118ea91424407f64b362d810ac5023d88a1 Mon Sep 17 00:00:00 2001
From: Tom <support@vamrs.com>
Date: Fri, 8 Jan 2021 04:13:06 +0800
Subject: [PATCH 19/38] drivers/net/phy/marvell: modify PHY LED config(VIC7100
 EVB)

---
 drivers/net/phy/marvell.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/net/phy/marvell.c b/drivers/net/phy/marvell.c
index afd44db2417d..65855aa5351a 100644
--- a/drivers/net/phy/marvell.c
+++ b/drivers/net/phy/marvell.c
@@ -777,6 +777,18 @@ static void marvell_config_led(struct phy_device *phydev)
 		return;
 	}
 
+#ifdef CONFIG_SOC_STARFIVE_VIC7100
+	/*
+	 * PHY(MARVELL_PHY_ID_88E1510) LED config:
+	 * Reserved.. 1000, Rst
+	 * LED[2] .. 0000,Link on
+	 * LED[1] .. 0001,Link on, Blink - Activity
+	 * LED[0] .. 0100,Blink - Activity
+	 */
+#define MII_88E1512_PHY_LED1_LINK_LED0_ACTIVE	0x1014
+	def_config = MII_88E1512_PHY_LED1_LINK_LED0_ACTIVE;
+#endif
+
 	err = phy_write_paged(phydev, MII_MARVELL_LED_PAGE, MII_PHY_LED_CTRL,
 			      def_config);
 	if (err < 0)
-- 
2.30.0

